akka-stream-testkit/src/main/scala/akka/stream/testkit/StreamTestKit.scala

```diff
@@ -0,0 +1,443 @@
+
+    def within[T](max: FiniteDuration)(f: ⇒ T): T = probe.within(0.seconds, max)(f)
+
+    def onSubscribe(subscription: Subscription): Unit = probe.ref ! OnSubscribe(subscription)
+    def onNext(element: I): Unit = probe.ref ! OnNext(element)
+    def onComplete(): Unit = probe.ref ! OnComplete
+    def onError(cause: Throwable): Unit = probe.ref ! OnError(cause)
+  }
+
+  /**
+   * Single subscription tracking for [[ManualProbe]].
+   */
+  class Probe[T] private[TestSubscriber] ()(implicit system: ActorSystem) extends ManualProbe[T] {
+
+    override type Self = Probe[T]
+
+    private lazy val subscription = expectSubscription()
+
+    def request(n: Long): Self = {
+      subscription.request(n)
+      this
+    }
+
+    def requestNext(element: T): Self = {
+      subscription.request(1)
+      expectNext(element)
+      this
+    }
+
+    def cancel(): Self = {
+      subscription.cancel()
+      this
+    }
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[testkit] object StreamTestKit {
+
+  import TestPublisher._
+  import TestSubscriber._
+
+  sealed trait PublisherEvent extends DeadLetterSuppression
+  final case class Subscribe(subscription: Subscription) extends PublisherEvent
+  final case class CancelSubscription(subscription: Subscription) extends PublisherEvent
+  final case class RequestMore(subscription: Subscription, elements: Long) extends PublisherEvent
+
+  sealed trait SubscriberEvent extends DeadLetterSuppression
+  final case class OnSubscribe(subscription: Subscription) extends SubscriberEvent
+  final case class OnNext[I](element: I) extends SubscriberEvent
+  final case object OnComplete extends SubscriberEvent
+  final case class OnError(cause: Throwable) extends SubscriberEvent
+
+  final case class CompletedSubscription[T](subscriber: Subscriber[T]) extends Subscription {
+    override def request(elements: Long): Unit = subscriber.onComplete()
+    override def cancel(): Unit = ()
+  }
+
+  final case class FailedSubscription[T](subscriber: Subscriber[T], cause: Throwable) extends Subscription {
+    override def request(elements: Long): Unit = subscriber.onError(cause)
+    override def cancel(): Unit = ()
+  }
+
+  final case class PublisherProbeSubscription[I](subscriber: Subscriber[_ >: I], publisherProbe: TestProbe) extends Subscription {
+    def request(elements: Long): Unit = publisherProbe.ref ! RequestMore(this, elements)
+    def cancel(): Unit = publisherProbe.ref ! CancelSubscription(this)
+
+    def expectRequest(n: Long): Unit = publisherProbe.expectMsg(RequestMore(this, n))
+    def expectRequest(): Long = publisherProbe.expectMsgPF() {
+      case RequestMore(sub, n) if sub eq this ⇒ n
+    }
+
+    def expectCancellation(): Unit = publisherProbe.fishForMessage() {
+      case CancelSubscription(sub) if sub eq this ⇒ true
+      case RequestMore(sub, _) if sub eq this     ⇒ false
+    }
+
+    def sendNext(element: I): Unit = subscriber.onNext(element)
```

Comment:
Not for this PR, but create a ticket if you like the idea:
We could track demand and throw exception if this is invoked when there is no demand.
We could have a power user flag that doesn't do that, to be able to simulate an invalid publisher.